### PR TITLE
Optimize logic and refactor code

### PR DESCRIPTION
**Pull request by jianxingxuejian**
## Description

Please carefully read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md) before making any pull requests.
And, **Do not make a pull request to merge into stable unless it is a hotfix. Use the development branch instead.**
## Type of changes

- [ ] Bug fix
- [ ] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

**Comment by Gmskjunior** [comment]
Exit

Em qui., 25 de ago. de 2022 11:41, null ***@***.***> escreveu:

File: src/main/java/emu/grasscutter/auth/DefaultAuthenticators.java [hunk 1]
```diff
@@ -29,222 +32,170 @@ public final class DefaultAuthenticators {
...
-                response.message = responseMessage;
-
+                if (account.getPassword() == null || account.getPassword().isEmpty()) {
+                    getLogger().info(translate("messages.dispatch.account.login_password_error", address));
+                    response.message = translate("messages.dispatch.account.password_length_error");
+                    return response;
+                }
+                if (!BCrypt.verifyer().verify(password.toCharArray(), account.getPassword()).verified) {
+                    getLogger().info(translate("messages.dispatch.account.login_password_error", address));
+                    response.message = translate("messages.dispatch.account.password_error");
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/auth/DefaultAuthenticators.java]:
```suggestion
                    response.message = translate("messages.dispatch.account.password_error");
                    return response;
```

**Review by Birdulon — CHANGES_REQUESTED** [review]

**Comment by Birdulon** [comment]
There's still a fair bit of duplication between `PasswordAuthenticator` and `ExperimentalPasswordAuthenticator`, it might be best to restructure `PasswordAuthenticator` to handle most of the steps in a new method, have `ExperimentalPasswordAuthenticator` inherit `PasswordAuthenticator` and just override a new method which does the actual password checking.

**Comment by jianxingxuejian** [comment]
yes, there is still a lot of room for optimization. I merge into one class, will it cause harm in the future?

**Comment by Birdulon** [comment]
> I merge into one class, will it cause harm in the future?

Plugins like to subclass/override the authenticators. I'd suggest leaving them in separate classes, but have `ExperimentalPasswordAuthenticator` subclass `PasswordAuthenticator`.

File: src/main/java/emu/grasscutter/server/http/objects/AuthenticateResponse.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package emu.grasscutter.server.http.objects;
+
+public class AuthenticateResponse {
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/server/http/objects/AuthenticateResponse.java]: small pet peeve: please rename to `AuthenticationResponse`

**Review by KingRainbow44 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**